pkg/alicloud/client/client_dns.go

```diff
@@ -119,6 +131,41 @@ func (d *dnsClient) DeleteDomainRecords(ctx context.Context, domainName, name, r
 	return nil
 }
 
+func (d *dnsClient) getDomainsWithCache() (map[string]alidns.Domain, error) {
+	if v, ok := d.domainsCache.Get(d.accessKeyID); ok {
+		return v.(map[string]alidns.Domain), nil
+	}
+	domains, err := d.getDomains()
+	if err != nil {
+		return nil, err
+	}
+	d.domainsCache.Set(d.accessKeyID, domains, domainsCacheTTL)
```

Comment:
It still has issue when a lot of dns records  with the same accessKeyId are reconciled. This can happen because `d.getDomains()` takes long time enough. As a result, lots of `dnsClient` will call `d.getDomains()` inevitably though most of them are expected to go through the cache.

I would suggest introduce extra locks:
1, maintain a globla lock table accessKeyIDLockTable map[accessKeyID][Lock]
2, a global lock: gCacheLock
3. Check domain cache with accessKeyID, if it doesn't exist, go to step 4.
4. retreive lock from accessKeyIDLockTable, if doesn't exist,go to 5. Otherwise, goto 6. The lock for accesKeyID level is accessKeyIDLock.
5. gCacheLock.Lock(),  put a lock in accessKeyIDLock  for the specified accessKeyID, gCacheLock.UnLock(). 
6. accessKeyIDLock.lock and fill with domain cache.

Reply:
Thanks for spotting this, you are right this can happen, and indeed we need additional locking to prevent it. But I don't think we need so many locks - a single lock on the (shared) cache should be enough, to ensure only a single `d.getDomains()` is performed. `cache.Expring` itself is already concurrency safe, we don't need to do anything special to protect its entries.

Reply:
Done, PTAL.